File: 프로그래머스/3/43164. 여행경로/여행경로.py
import heapq

def dfs(visited, start, dic, result, n):
    if len(result) == n + 1:
        return True
    
    for end in sorted(dic.get(start, [])):
        if visited[(start, end)] == 0:
            continue
            
        visited[(start, end)] -= 1
        result.append(end)
        if dfs(visited, end, dic, result, n):
            return True
        result.pop()
        visited[(start, end)] += 1
    

def solution(tickets):
    dic = {}
    visited = {}
    for a, b in tickets:
        dic[a] = dic.get(a, [])
        dic[a].append(b)
        visited[(a, b)] = visited.get((a, b), 0) + 1
        
    result = ["ICN"]
        
    dfs(visited, "ICN", dic, result, len(tickets))
    
    return result
        

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.stream.*;
import java.util.*;
class Solution {
    Map<String[], Integer> visited = new HashMap<>();
    Map<String, List<String[]>> go = new HashMap<>();
    LinkedList<String> ls = new LinkedList<>();
    List<String[]> result = new ArrayList<>();
    public String[] solution(String[][] tickets) {
        for(String[] m : tickets){
            visited.put(m, 0);
            
            List<String[]> gogo = go.getOrDefault(m[0],new ArrayList<>());
            gogo.add(m);
            go.put(m[0], gogo);
        }
        ls.addLast("ICN");
        dfs();
        
        return result.stream()
            .sorted(Comparator.comparing(s -> String.join("",s)))
            .findFirst().get();
    }
    
    public void dfs(){
        if (ls.size() - 1 == visited.size()){
            result.add(ls.stream().toArray(String[]::new));
            return;
        }
        
        for(String[] s: go.getOrDefault(ls.peekLast(), new ArrayList<>())){
            if (visited.get(s) == 1) continue;
            
            visited.put(s, 1);
            ls.addLast(s[1]);
            dfs();
            ls.pollLast();
            visited.put(s, 0);
        }
    }
}